src/coreclr/tools/aot/ILCompiler.ReadyToRun/JitInterface/CorInfoImpl.ReadyToRun.cs

```diff
@@ -2945,6 +2926,16 @@ private void reportInliningDecision(CORINFO_METHOD_STRUCT_* inlinerHnd, CORINFO_
                 // If we didn't succeed on the inline, just pop back to the state before inlining
                 _precodeFixups = _stashedPrecodeFixups.Pop();
                 _inlinedMethods = _stashedInlinedMethods.Pop();
+                if (inlineResult == CorInfoInline.INLINE_FAIL && inlineeHnd != null)
+                {
+                    MethodDesc failedInlinee = HandleToObject(inlineeHnd);
+                    if (failedInlinee.IsPInvoke)
+                    {
+                        // If we failed to inline a P/Invoke, then we need to request compilation for it
+                        // to avoid generating a stub at runtime.
+                        _compilation.PrepareForCompilationRetry(_compilation.NodeFactory.CompiledMethodNode(failedInlinee), Array.Empty<EcmaMethod>());
```

Comment:
I don't believe this is the correct approach, we should use the standard dependencies mechanism for this, but it looks like we might not have the correct hooks for this at the moment. Contact me on Monday and I'll look into this with you.

Reply:
We have `_additionalDependencies` in CorInfoImpl.cs that is used outside READYTORUN. It might be the thing to move out of `if READYTORUN` (and adding the respective `InitializeNonRelocationDependencies`) to solve this.